src/julia_to_gap.jl

```diff
@@ -142,66 +152,64 @@ function julia_to_gap(
 end
 
 ## Convert two dimensional arrays
-function julia_to_gap(
+function julia_to_gap_internal(
     obj::Matrix{T},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
+    recursion_dict::GapCacheDict,
+    recursive::Bool,
 ) where {T}
-    (rows, cols) = size(obj)
-    if haskey(recursion_dict, obj)
-        return recursion_dict[obj]
+    if recursion_dict !== nothing && haskey(recursion_dict, obj)
+      return recursion_dict[obj]
     end
+    (rows, cols) = size(obj)
     ret_val = NewPlist(rows)
     if recursive
+        if recursion_dict === nothing
+          recursion_dict = RecDict()
+        end
         recursion_dict[obj] = ret_val
     end
     for i = 1:rows
-        ret_val[i] = julia_to_gap(obj[i, :], recursion_dict; recursive)
+        ret_val[i] = julia_to_gap_internal(obj[i, :], recursion_dict, recursive)
```

Comment:
Existing issue, but shouldn't this also do the `get!(recursion_dict::RecDict, x)` dance?

Reply:
I think no.
The idea of the `get!` call is twofold.
1. Check whether the object in question is already in the dictionary, and if yes then take the value.
2. If the object is not yet in the dictionary then store it together with the converted value.
Situation 1. cannot occur because we are just now creating the Julia object anew.
And situation 2. would not help later on because no pointer to this newly created object can be found in the recursion.
(I will add a comment.)

Reply:
I feel there is something inconsistent here. Because then the object produced by recurively called `julia_to_gap_internal` is potentially never added to `recursion_dict`, is it? So if we had `T = BigInt ` we would not be caching the conversion results for numerator or denominator. While if we convert a list of `BigInt` they would be cached.

Perhaps we never want to cache a `BigInt`'s converted while. But then why does the `julia_to_gap_internal` method for `Rational{T}` provide the the full 3-argument form instead of just using `GAP.@install`?

Reply:
OK the crucial point I missed is that we convert row-wise, i.e. we convert `obj[i, :]` and not elementwise, i.e. `obj[i,j]`